fastlane_bot/helpers/txhelpers.py

```diff
@@ -117,578 +54,150 @@ def validate_and_submit_transaction(
+                f"Most often this is due to the arb opportunity already being closed\n."
+                f"This is expected to happen occasionally, discarding; exception: {e}"
+            )
             return None
-        else:
-            access_list = loads(response.text)["result"]["accessList"]
-            return access_list
 
-    def construct_contract_function(
-        self,
-        routes: List[Dict[str, Any]],
-        src_amt: int,
-        src_address: str,
-        flashloan_struct=None,
-    ):
-        """
-        Builds the transaction using the Arb Contract function. This version can generate transactions using flashloanAndArb and flashloanAndArbV2.
+        if self.access_lists:
+            try:
+                access_list = self.eth.create_access_list({"from": self.wallet_address, "to": self.arb_contract.address, "data": tx["data"]})
+                tx_after = {**tx, "accessList": access_list}
+                estimated_gas_after = self.eth.estimate_gas(transaction=tx_after)
+                if estimated_gas > estimated_gas_after:
+                    estimated_gas = estimated_gas_after
+                    tx = tx_after
+                self.cfg.logger.info(f"Access list: {access_list}\n, gas before and after: {estimated_gas}, {estimated_gas_after}")
+            except Exception as e:
+                self.cfg.logger.info(f"Applying access list to transaction failed with {e}")
 
-        routes: the routes to be used in the transaction
-        src_amt: the amount of the source token to be sent to the transaction
-        gas_price: the gas price to be used in the transaction
+        tx["gas"] = estimated_gas + self.cfg.DEFAULT_GAS_SAFETY_OFFSET
 
-        returns: the transaction function ready to be submitted
-        """
-        if self.ConfigObj.SELF_FUND:
-            return self.arb_contract.functions.fundAndArb(
-                routes, src_address, src_amt
-            )
-        return self.arb_contract.functions.flashloanAndArbV2(
-            flashloan_struct, routes
-        )
+        raw_tx = self.eth.account.sign_transaction(tx, self.cfg.ETH_PRIVATE_KEY_BE_CAREFUL).rawTransaction
 
-    def _handle_function_build_exception(self, exception: Exception) -> int or None:
-        """Handles exceptions that occur during transaction building.
+        gas_cost_wei = int(gas_price * estimated_gas * self.cfg.EXPECTED_GAS_MODIFIER)
+        if self.cfg.network.GAS_ORACLE_ADDRESS:
+            gas_cost_wei += self.cfg.GAS_ORACLE_CONTRACT.caller.getL1Fee(raw_tx)
 
-        This method logs the exception and attempts to extract information from it. If the exception
-        indicates that the maximum fee per gas is less than the block base fee, it extracts and returns
-        the base fee. Otherwise, it logs a warning and returns None.
+        gas_cost_eth = Decimal(gas_cost_wei) / ETH_DECIMALS
+        gas_cost_usd = gas_cost_eth * expected_profit_usd / expected_profit_gastkn
```

Comment:
that's ugly -- why don't you have as parameter a expected profit figure and an exchange rate, rather than having to expected profit figures that are then used to back out the exchange rate?

Reply:
I kept the previous implementation here.

I will check to see if the caller may as well pass already-processed input (BTW, I have done that for a bunch of other input parameters, sometimes even removing them altogether, but not for this one).